package com.playtube.protube.video.music.local.subscription.workers

import java.io.InputStream
import java.io.OutputStream
import kotlinx.serialization.ExperimentalSerializationApi
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.decodeFromStream
import kotlinx.serialization.json.encodeToStream
import org.schabi.newpipe.extractor.subscription.SubscriptionExtractor.InvalidSourceException

/**
 * A JSON implementation capable of importing and exporting subscriptions, it has the advantage
 * of being able to transfer subscriptions to any device.
 */
object ImportExportJsonHelper {
    private val json = Json { encodeDefaults = true }

    /**
     * Read a JSON source through the input stream.
     *
     * @param in            the input stream (e.g. a file)
     * @return the parsed subscription items
     */
    @JvmStatic
    @Throws(InvalidSourceException::class)
    fun readFrom(`in`: InputStream?): List<SubscriptionItem> {
        if (`in` == null) {
            throw InvalidSourceException("input is null")
        }

        try {
            @OptIn(ExperimentalSerializationApi::class)
            return json.decodeFromStream<SubscriptionData>(`in`).subscriptions
        } catch (e: Throwable) {
            throw InvalidSourceException("Couldn't parse json", e)
        }
    }

    /**
     * Write the subscriptions items list as JSON to the output.
     *
     * @param items         the list of subscriptions items
     * @param out           the output stream (e.g. a file)
     */
    @OptIn(ExperimentalSerializationApi::class)
    @JvmStatic
    fun writeTo(
        items: List<SubscriptionItem>,
        out: OutputStream
    ) {
        json.encodeToStream(SubscriptionData(items), out)
    }
}
